### PR TITLE
Make impossible to extend a lock using lock with different

### PR DESCRIPTION
**Pull request by AndrewPashkin**
Related to #25

### CODE REVIEW

**Comment by AndrewPashkin** [comment]
I've changed the exception type raised from `extend()`, but now I have doubts about it.

**Comment by ionelmc** [comment]
What if you raise `NotAcquired` instead? It's more consistent with the actual check.

**Comment by AndrewPashkin** [comment]
The main point is that according to the Redis [docs for TTL][1]:

> In Redis 2.6 or older the command returns -1 if the key does not exist or if the key exist but has no associated expire.
>Starting with Redis 2.8 the return value in case of error changed:
>
>    - The command returns -2 if the key does not exist.
>    - The command returns -1 if the key exists but has no associated expire.

So the error might be raised not only in case, when a key has no expiration time, but also when it doesn't exist.

[1]: http://redis.io/commands/TTL

**Comment by AndrewPashkin** [comment]
And regarding error thrown when no `expire` arguments provided for `extend()` and/or `__init__()` - I don't think `NotExpireable` fits for that case, because it's more a programming error or type error (inconsistent arguments), rather than absence of possibility of setting expiration for a key.

**Comment by ionelmc** [comment]
>  I don't think NotExpireable fits for that case,

I suggested `NotAcquired` ... or do you have a different situation in mind?

**Comment by ionelmc** [comment]
> So the error might be raised not only in case, when a key has no expiration time, but also when it doesn't exist.

I think this is fine, since the script is atomic you can do a check first, for the existence of the key. It would be a tiny bit slower but it ain't that bad. I don't think anyone would be able to tell the speed difference :-)

**Comment by AndrewPashkin** [comment]
> I suggested NotAcquired ... or do you have a different situation in mind?
Problem here, that `TTL` distinguishes two errors only starting from 2.8 version of Redis. And the library supports Redis 2.6.12. Probably checking for the case, when a key is absent should be performed manually.

**Comment by AndrewPashkin** [comment]
> I think this is fine, since the script is atomic you can do a check first, for the existence of the key. It would be a tiny bit slower but it ain't that bad. I don't think anyone would be able to tell the speed difference :-)

I agree on that.

But what about absence of `expire` argument? `NotAcquired` doesn't fits there.

**Comment by ionelmc** [comment]
In that case it makes sense to have the old `NotExpirable` exception. We could make the `NotExpirable` subclass `TypeError`.

**Comment by AndrewPashkin** [comment]
In my mind `NotExpireable` is for the case, when the lock doesn't have a TTL assigned initially in Redis, and when user didn't provided `expire` argument and wants to extend expiration - is a programming error. In a moment when this error occurs - there is no knowledge of whether is lock is expireable or not.

**Comment by ionelmc** [comment]
Well then subclassing `RuntimeError` seems fine no?

**Comment by AndrewPashkin** [comment]
> Well then subclassing RuntimeError seems fine no?

https://docs.python.org/2/library/exceptions.html#exceptions.TypeError
> exception RuntimeError

> Raised when an error is detected that doesn’t fall in any of the other categories. The associated value is a string indicating what precisely went wrong.

https://docs.python.org/2/library/exceptions.html#exceptions.RuntimeError

> exception TypeError

> Raised when an operation or function is applied to an object of inappropriate type. The associated value is a string giving details about the type mismatch.
https://docs.python.org/2/library/exceptions.html#exceptions.TypeError

I think, that `TypeError` fits better. Correct me if I'm wrong - seems like it's a standard way in Python to throw a `TypeError` in case of some inconsistency in passed arguments, for example if you call `min()` without arguments, it will throw `TypeError`.

And also - why subclassing at all? I think, `TypeError` is exactly for such cases.

**Comment by ionelmc** [comment]
> And also - why subclassing at all? I think, TypeError is exactly for such cases.

I don't like people to check stuff inside the exception instance to figure out what the problem actually is. Same reason Python 3 has specific OSError subclasses instead of the nasty generic error you have on Python 2 (where you have to check `.errno` all over the place).

**Comment by AndrewPashkin** [comment]
What about `min()` and how it raises `TypeError` when no arguments provided? I think it's closer to our case.

**Comment by ionelmc** [comment]
The questions you got to ask yourself are:

* is it specific enough?
* can there be other situations where exc of same type would be raise?

I think it's possible to get a TypeError if you try to make a lock on a broken connection object. And that would make raising TypeError for this argument issue not specific enough. Color me crazy and paranoid :grin:

**Comment by AndrewPashkin** [comment]
What we have in this case exactly the Type Error situation - user provided nonsense arguments trying to extend a lock without passing extension time.

**Comment by ionelmc** [comment]
Ok, fine. Lets use a `TypeError` and see how it goes.

**Comment by AndrewPashkin** [comment]
By the way, in that PR tests are failing too, do you have ideas why? Maybe Travis is having some trouble in launching 125 concurrent processes?

**Comment by ionelmc** [comment]
Well you wrote the test, you tell me :)

PyPy can be quite slow in some situations. 
Maybe ` time.sleep(0.5)` needs to wait more in test_overlap2?

File: src/redis_lock/__init__.py [hunk 1]
```diff
@@ -195,12 +213,22 @@ def extend(self, expire=None):
...
             lock initialization will be taken.
         """
-        if self._expire is None:
-            raise NotExpirable('The lock has no expiry time, so extending it '
-                               'makes no sense.')
-        if expire is None:
+        if expire is None and self._expire is not None:
             expire = self._expire
-        self._client.set(self._name, self._id, xx=True, ex=expire)
+        elif expire is None and self._expire is None:
```
Comment (ionelmc) [inline src/redis_lock/__init__.py]: You can check for `expire is None` just once I think. It's one indentation level more but then it's clear what's the common part.

**Comment by ionelmc** [comment]
This looks safeish to merge (the pypy is just too slow sometimes - ok to fail).

You can undo the tox version pin btw, 2.3.1 (released couple days ago) fix the config loading issue.

**Comment by ionelmc** [comment]
Actually :grin:

**Comment by AndrewPashkin** [comment]
Noooo, I just didn't have enough time to sort this out!

**Comment by ionelmc** [comment]
No worries, you can make new PR :)

**Pull request merged**